storm-server/src/main/java/org/apache/storm/scheduler/resource/normalization/NormalizedResources.java

```diff
@@ -384,14 +391,29 @@ public double calculateMinPercentageUsedBy(NormalizedResources used, double tota
 
     /**
      * If a node or rack has a kind of resource not in a request, make that resource negative so when sorting that node or rack will
-     * be less likely to be selected.
+     * be less likely to be selected. If the resource is in the request, make that resource positive.
      * @param request the requested resources.
      */
     public void updateForRareResourceAffinity(NormalizedResources request) {
         int length = Math.min(this.otherResources.length, request.otherResources.length);
         for (int i = 0; i < length; i++) {
-            if (request.getResourceAt(i) == 0.0) {
-                this.otherResources[i] = -1 * this.otherResources[i];
+            if (request.getResourceAt(i) == 0.0 && this.otherResources[i] > 0.0) {
+                this.otherResources[i] = -this.otherResources[i]; // make negative
+            } else if (request.getResourceAt(i) > 0.0 && this.otherResources[i] < 0.0) {
```

Comment:
The goal in this code is to make unrequested resource negative (but also requested resource positive). So when processing a new executor with different resource requirements, it can be reapplied and sort resulting elements resorted. With earlier code, this was a "use-once-only" method on the "otherResource". With the change above, the object can be reused.

Reply:
Sounds good. But since the object (otherResource) is not reused currently, there is no bug found. Is it correct?

Reply:
I am not sure this has been reported or not.

Reply:
But I ran into it when I was trying to optimize the sorting code with reused objects. across executors (you may remember the executor caching key that I removed below getExecRequestKey). I will attempt that speed in a different Jira. But this method needs to work properly and repeatedly with two or more calls.

Reply:
Thanks